### DISCUSSION TITLE
Enhancing UC Opening Speed

### DISCUSSION THREAD

**Discussion by lucasmarquex** [seleniumbase/SeleniumBase#2536] [Q&A] [answered]
Before asking my question, I'd like to congratulate you on your work with Undetected Chromedriver (UC), Michael. It has proven to be much more stealthy than the UC from Ultrafunk Amsterdam.

My question is: Chrome, when used with SeleniumBase, takes about 3 seconds to open, whereas Undetected Chromedriver opens in 0.5 seconds. Is there any way to speed up the opening process? I can only open one window at a time using UC due to conflicts, so I'm using threading.Lock, and this significantly impacts the use of multiple windows.

Is there any way to disable all seleniumbase features except UC and make it open faster? Or determine the port so there is no conflict and you can open several windows simultaneously?

**Comment by mdmintz** [comment, marked as answer]
Hello! You'll find your answer in the `driver`-specific methods added during UC Mode:

```python
driver.uc_open(url)
driver.uc_open_with_tab(url)
driver.uc_open_with_reconnect(url, reconnect_time=None)
driver.reconnect(timeout)
driver.uc_click(selector)
driver.uc_switch_to_frame(frame)
```

Use `driver.uc_open_with_reconnect(url, reconnect_time)` with a short `reconnect_time`, eg. `0.5` seconds. It's set to around 3 seconds so that there's enough time to get past a Cloudflare verification check before Selenium reconnects to the browser, which makes it detectable.

If you want to open a webpage with no delay at all while using UC Mode, then use `driver.default_get(url)`. That will make you detectable, but it might be that you don't always need to avoid detection when opening web pages.

For multithreading, use the `pytest` format, which has a special option for that. More info: https://github.com/seleniumbase/SeleniumBase/issues/2006#issuecomment-1679789785

**Comment by lucasmarquex** [comment]
I am referring that there is this delay in opening the browser (chrome.exe process), with seleniumbase, opening becomes slow

**Reply by mdmintz** [reply]
If it's the initial browser open after downloading `chromedriver` (which gets renamed as `uc_driver` in UC Mode) then there is an expected pause for `uc_driver` getting modified so that the driver no longer makes Selenium detectable in the browser.

**Comment by lucasmarquex** [comment]
Even without seleniumbase's UC mode and with the chrome driver already installed, it usually takes 4x longer than undetected_chromedriver, some initial check is generating a lot of extra time to open the browser process with chromedriver

**Reply by mdmintz** [reply]
Could you identify the bottleneck for me? I haven't seen it, but maybe it's specific to your OS. Put a `breakpoint()` in your code and see if you can trace where the issue is originating. If there's a real issue, I may be able to fix it.

**Comment by lucasmarquex** [comment]
Test:
![image](https://github.com/seleniumbase/SeleniumBase/assets/41024064/e092676c-7fb1-439c-85b5-ae0678cc0d85)

Result:
![image](https://github.com/seleniumbase/SeleniumBase/assets/41024064/d40cdfe9-cba9-402d-a0e4-609de1ac27f8)

**Reply by mdmintz** [reply]
It should be `Driver(uc=True)` ... not `Driver(uc=uc)` (where your `uc` is where you imported `undetcted_chromedriver` into.

**Reply by lucasmarquex** [reply]
Corrected, the result was the same

**Comment by lucasmarquex** [comment]
Does the interference problem have to do with using port 9222? Because it seems to be a difference between UC mode and undetected_chromedriver. And what is the difference between being undetectable between the two? The difference is that Seleniumbase's UC mode handles the created profile?

**Reply by mdmintz** [reply]
It'll use port 9222 if available. Otherwise it'll find another free port. The modes work very similarly.

Will need more info as to why you're seeing the slowdown. Will likely need `breakpoint()` and a specific line where the slowdown occurred.

**Reply by lucasmarquex** [reply]
What is the difference between the two libs? because seleniumbase ucmode has interference with multiple windows and undetected_chromedriver does not

**Reply by mdmintz** [reply]
SeleniumBase has a larger API, more features, more customization options, etc, but the general algorithm for avoiding detection is similar. SeleniumBase tweaks the algorithm a bit so that it can bypass detection on more websites.
